File: extension/src/main/java/io/holunda/camunda/bpm/data/factory/MapVariableFactory.java
package io.holunda.camunda.bpm.data.factory;

import io.holunda.camunda.bpm.data.adapter.ReadAdapter;
import io.holunda.camunda.bpm.data.adapter.WriteAdapter;
import io.holunda.camunda.bpm.data.adapter.map.MapReadWriteAdapterRuntimeService;
import io.holunda.camunda.bpm.data.adapter.map.MapReadWriteAdapterTaskService;
import io.holunda.camunda.bpm.data.adapter.map.MapReadWriteAdapterVariableMap;
import io.holunda.camunda.bpm.data.adapter.map.MapReadWriteAdapterVariableScope;
import org.camunda.bpm.engine.RuntimeService;
import org.camunda.bpm.engine.TaskService;
import org.camunda.bpm.engine.delegate.VariableScope;
import org.camunda.bpm.engine.variable.VariableMap;
import org.jetbrains.annotations.NotNull;

import java.util.Map;
import java.util.Objects;

/**
 * Variable factory of a base parametrized map type.
 *
 * @param <K> member key type of the factory.
 * @param <V> member value type of the factory.
 */
public class MapVariableFactory<K, V> implements VariableFactory<Map<K, V>> {

  @NotNull
  private final String name;

  @NotNull
  private final Class<K> keyClazz;

  @NotNull
  private final Class<V> valueClazz;

  public MapVariableFactory(@NotNull String name, @NotNull Class<K> keyClazz, @NotNull Class<V> valueClazz) {
    this.name = name;
    this.keyClazz = keyClazz;
    this.valueClazz = valueClazz;
  }

  @Override
  public WriteAdapter<Map<K, V>> on(VariableScope variableScope) {
    return new MapReadWriteAdapterVariableScope<>(variableScope, name, keyClazz, valueClazz);
  }

  @Override
  public ReadAdapter<Map<K, V>> from(VariableScope variableScope) {
    return new MapReadWriteAdapterVariableScope<>(variableScope, name, keyClazz, valueClazz);
  }

  @Override
  public WriteAdapter<Map<K, V>> on(VariableMap variableMap) {
    return new MapReadWriteAdapterVariableMap<>(variableMap, name, keyClazz, valueClazz);
  }

  @Override
  public ReadAdapter<Map<K, V>> from(VariableMap variableMap) {
    return new MapReadWriteAdapterVariableMap<>(variableMap, name, keyClazz, valueClazz);
  }

  @Override
  public WriteAdapter<Map<K, V>> on(RuntimeService runtimeService, String executionId) {
    return new MapReadWriteAdapterRuntimeService<>(runtimeService, executionId, name, keyClazz, valueClazz);
  }

  @Override
  public ReadAdapter<Map<K, V>> from(RuntimeService runtimeService, String executionId) {
    return new MapReadWriteAdapterRuntimeService<>(runtimeService, executionId, name, keyClazz, valueClazz);
  }

  @Override
  public WriteAdapter<Map<K, V>> on(TaskService taskService, String taskId) {
    return new MapReadWriteAdapterTaskService<>(taskService, taskId, name, keyClazz, valueClazz);
  }

  @Override
  public ReadAdapter<Map<K, V>> from(TaskService taskService, String taskId) {
    return new MapReadWriteAdapterTaskService<>(taskService, taskId, name, keyClazz, valueClazz);
  }

  @Override
  @NotNull
  public String getName() {
    return name;
  }

  /**
   * Retrieves key type.
   *
   * @return key type.
   */
  @NotNull
  public Class<K> getKeyClass() {
    return keyClazz;
  }

  /**
   * Retrieves value type.
   *
   * @return value type.
   */
  @NotNull
  public Class<V> getValueClass() {
    return valueClazz;
  }

  @Override public boolean equals(Object o) {
    if (this == o)
      return true;
    if (o == null || getClass() != o.getClass())
      return false;
    MapVariableFactory<?, ?> that = (MapVariableFactory<?, ?>) o;
    return name.equals(that.name) &&
      keyClazz.equals(that.keyClazz) &&
      valueClazz.equals(that.valueClazz);
  }

  @Override public int hashCode() {
    return Objects.hash(name, keyClazz, valueClazz);
  }
}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/factory/MapVariableFactory.kt
package io.holunda.camunda.bpm.data.factory

import io.holunda.camunda.bpm.data.adapter.ReadAdapter
import io.holunda.camunda.bpm.data.adapter.WriteAdapter
import io.holunda.camunda.bpm.data.adapter.map.*
import org.camunda.bpm.engine.CaseService
import org.camunda.bpm.engine.RuntimeService
import org.camunda.bpm.engine.TaskService
import org.camunda.bpm.engine.delegate.VariableScope
import org.camunda.bpm.engine.externaltask.LockedExternalTask
import org.camunda.bpm.engine.variable.VariableMap
import java.util.*

/**
 * Variable factory of a base parametrized map type.
 *
 * @param [K]> member key type of the factory.
 * @param [V] member value type of the factory.
 */
class MapVariableFactory<K, V>(
  override val name: String,
  val keyClass: Class<K>,
  val valueClass: Class<V>
) : VariableFactory<Map<K, V>> {

  companion object {
    inline fun <reified K, reified V> forType(name: String) = MapVariableFactory(name, K::class.java, V::class.java)
  }

  override fun on(variableScope: VariableScope): WriteAdapter<Map<K, V>> {
    return MapReadWriteAdapterVariableScope(
      variableScope,
      name,
      keyClass,
      valueClass
    )
  }

  override fun from(variableScope: VariableScope): ReadAdapter<Map<K, V>> {
    return MapReadWriteAdapterVariableScope(
      variableScope,
      name,
      keyClass,
      valueClass
    )
  }

  override fun on(variableMap: VariableMap): WriteAdapter<Map<K, V>> {
    return MapReadWriteAdapterVariableMap(
      variableMap,
      name,
      keyClass,
      valueClass
    )
  }

  override fun from(variableMap: VariableMap): ReadAdapter<Map<K, V>> {
    return MapReadWriteAdapterVariableMap(
      variableMap,
      name,
      keyClass,
      valueClass
    )
  }

  override fun on(runtimeService: RuntimeService, executionId: String): WriteAdapter<Map<K, V>> {
    return MapReadWriteAdapterRuntimeService(
      runtimeService,
      executionId,
      name,
      keyClass,
      valueClass
    )
  }

  override fun from(runtimeService: RuntimeService, executionId: String): ReadAdapter<Map<K, V>> {
    return MapReadWriteAdapterRuntimeService(
      runtimeService,
      executionId,
      name,
      keyClass,
      valueClass
    )
  }

  override fun on(taskService: TaskService, taskId: String): WriteAdapter<Map<K, V>> {
    return MapReadWriteAdapterTaskService(
      taskService,
      taskId,
      name,
      keyClass,
      valueClass
    )
  }

  override fun from(taskService: TaskService, taskId: String): ReadAdapter<Map<K, V>> {
    return MapReadWriteAdapterTaskService(
      taskService,
      taskId,
      name,
      keyClass,
      valueClass
    )
  }

  override fun on(caseService: CaseService, caseExecutionId: String): WriteAdapter<Map<K, V>> {
    return MapReadWriteAdapterCaseService(
      caseService,
      caseExecutionId,
      name,
      keyClass,
      valueClass
    )
  }

  override fun from(caseService: CaseService, caseExecutionId: String): ReadAdapter<Map<K, V>> {
    return MapReadWriteAdapterCaseService(
      caseService,
      caseExecutionId,
      name,
      keyClass,
      valueClass
    )
  }

  override fun from(lockedExternalTask: LockedExternalTask): ReadAdapter<Map<K, V>> {
    return MapReadAdapterLockedExternalTask(
      lockedExternalTask,
      name,
      keyClass,
      valueClass
    )
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val that = other as MapVariableFactory<*, *>
    return name == that.name && keyClass == that.keyClass && valueClass == that.valueClass
  }

  override fun hashCode(): Int {
    return Objects.hash(name, keyClass, valueClass)
  }

  override fun toString(): String {
    return "MapVariableFactory{" +
      "name='" + name + '\'' +
      ", keyClazz=" + keyClass +
      ", valueClazz=" + valueClass +
      '}'
  }
}
